larq_compute_engine/core/bgemm_kernels_arm64.h

```diff
@@ -819,76 +836,192 @@ void BinaryKernelNeonOutOfOrder8x4(
       "mov %[lhs_ptr], %[lhs_col_ptr]\n"
       "mov %[rhs_ptr], %[rhs_col_ptr]\n"
 
+      // Compute how much of the 8x4 block of destination values that we have
+      // computed fits in the destination matrix. Typically, all of it fits, but
+      // when the destination matrix shape is not a multiple of 8x4 there are
+      // some 8x4 blocks along the boundaries that do not fit entirely.
+      "sub w1, %w[dst_rows], %w[row]\n"
+      "sub w2, %w[dst_cols], %w[col]\n"
+      "mov w3, #8\n"
+      "mov w4, #4\n"
+      "cmp w1, #8\n"
+      // Compute w1 = how many rows of the 8x4 block fit.
+      "csel w1, w1, w3, le\n"
+      "cmp w2, #4\n"
+      // Compute w2 = how many cols of the 8x4 block fit.
+      "csel w2, w2, w4, le\n"
+
+IF_BITPACKED_OUTPUT(
+      // Load the thresholds pointer, offset as needed by the current row, and
+      // load 8 values into v29 and v30.
+      "ldr x14, [%[params], #" RUY_STR(RUY_OFFSET_OUTPUT_TRANSFORM_THRESHOLDS) "]\n"
+      "add x14, x14, %x[row], lsl #2\n"
+      "ld1 {v29.4s, v30.4s}, [x14]\n"
+
+      // Use "unsigned extend long" instructions to extend to int32 (in order to
+      // do the comparison with the thresholds).
+      "uxtl v20.4s, v24.4h\n"
+      "uxtl2 v21.4s, v24.8h\n"
+      "uxtl v22.4s, v25.4h\n"
+      "uxtl2 v23.4s, v25.8h\n"
+      "ld1 {v8.4s}, [%[rhs_ptr]], #16\n" // In parallel, load next RHS data.
+      "uxtl v24.4s, v26.4h\n"
+      "uxtl2 v25.4s, v26.8h\n"
+      "uxtl v26.4s, v27.4h\n"
+      "uxtl2 v27.4s, v27.8h\n"
+
+      // We want to write a 1-bit if accum > threshold, and so we use "compare
+      // signed greater than" instructions, which fill each element with ones if
+      // accum > threshold, or zeroes otherwise.
+      "ld1 {v0.4s, v1.4s}, [%[lhs_ptr]], #32\n" // In parallel, load next LHS data.
+      "cmgt v20.4s, v20.4s, v29.4s\n"
+      "cmgt v21.4s, v21.4s, v30.4s\n"
+      "cmgt v22.4s, v22.4s, v29.4s\n"
+      "cmgt v23.4s, v23.4s, v30.4s\n"
+      "ld1 {v2.4s, v3.4s}, [%[lhs_ptr]], #32\n" // In parallel, load next LHS data.
+      "cmgt v24.4s, v24.4s, v29.4s\n"
+      "cmgt v25.4s, v25.4s, v30.4s\n"
+      "cmgt v26.4s, v26.4s, v29.4s\n"
+      "cmgt v27.4s, v27.4s, v30.4s\n"
```

Comment:
This is a bit ugly. Basically, the accumulators are *unsigned* uint16 (even though `AccumScalar` is signed int16†), but the thresholds have to be signed integers, and thus there's no way to correctly do this comparison at int16.

The thresholds have to be signed, because there has the be a threshold value that causes a 1-bit to *always* be written (e.g. because of a small multiplier and huge negative bias for a given channel. As we write a 1-bit if and only if `accum` > `threshold`, and accumulator values are >= 0, a threshold value that forces a 1-bit to be written in all circumstances must be strictly negative.

Thus, this code I'm introducing works by extending the int16 accumulators to int32, doing the comparison there, then narrowing the result againg and collecting the bits.

An alternative solution would be to make the accumulator type actually be int16, i.e. only use this faster kernel if the maximum accumulator value is <= 2^15 - 1. Then the comparison could correctly be done at int16 without extension, but at the cost of fewer input combinations being able to use this kernel as opposed to the fallback.

† It would probably make more sense to have the accumulator types be uint32 and uint16 respectively, since that's how we actually use them.